content/posts/2022-05-13-calculating-type-sets.markdown

```diff
@@ -0,0 +1,607 @@
+probably have to try out all possible input combinations to see if any of them
+does. That's easy for three arguments, but for \\(n\\) arguments there are
+\\(2^n\\) possible assignments, so it takes *exponential* time in the number of
+arguments.
+
+The problem of finding arguments that makes such a function return `true` (or
+proving that no such arguments exists) is called "boolean satisfiability" and
+it is NP-complete.
+
+It turns out to be extremely important *in what form* the expression is given,
+though. Some forms make it pretty easy to solve, while others make it hard.
+
+For example, every expression can be rewritten into what is called a
+[“Disjunctive Normal Form” (DNF)](https://en.wikipedia.org/wiki/Disjunctive_normal_form),
+so called because it consists of a series of *conjunction* (`&&`) terms, joined
+together by *disjunction* (`||`) operators[^3]:
+
+```go
+func F_DNF(x, y, z bool) bool {
+  return (x && z) || (!y && z)
+}
+```
+
+(You can verify that this is the same function as above, by
+[trying out all 8 input combinations](https://go.dev/play/p/dCtSs3tf91F))
+
+Each term has a subset of the arguments, possibly negated, joined by
+`&&`. The terms are then joined together using `||`.
+
+Solving the satisfiability problem for an expression in DNF is easy:
+
+1. Go through the individual terms. `||` is `true` if and only if
+   either of its operands is `true`, so for each term:
+   * If it contains both an argument and its negation (`x && !x`)
+     it can never be `true`, continue to the next term.
+   * Otherwise, you can infer valid arguments from the term:
+       - If it contains `x`, then we must pass `true` for `x`
+       - If it contains `!x`, then we must pass `false` for `x`
+       - If it contains neither, then what we pass for `x` does not matter and
+         either value works.
+   * The term then evaluates to `true` with these arguments, so the entire
+     expression does.
+2. If none of the terms can be made `true`, the function can never return
+   `true` and there is no valid set of arguments.
+
+On the other hand, there is also a [“Conjunctive Normal Form”
+(CNF)](https://en.wikipedia.org/wiki/Conjunctive_normal_form). Here, the
+expression is a series of *disjunction* (`||`) terms, joined together with
+*conjunction* (`&&`) operators:
+
+```go
+func F_CNF(x, y, z bool) bool {
+  return (!x || z) && (y || z) && (x || !y)
+}
+```
+
+(Again, you can [verify that this is the same function](https://go.dev/play/p/0xldLVGqu7m))
+
+For this, the idea of our algorithm does not work. To find a solution, you have
+to take *all terms* into account simultaneously. You can't just tackle them one
+by one. In fact, solving satisfiability on CNF (often abbreviated as “CNFSAT”)
+is NP-complete[^4].
+
+[It turns out](https://en.wikipedia.org/wiki/Functional_completeness) that
+*every* boolean function can be written as a single expression using only `||`, `&&` and `!`. In particular, every boolean has a DNF and a CNF.
```

Comment:
This should be “every boolean function”, right? The word “function” is missing.

Reply:
indeed.